conf4j-core/src/main/java/org/conf4j/core/ConfigurationProvider.java

```diff
@@ -0,0 +1,15 @@
+package org.conf4j.core;
+
+import java.util.function.BiConsumer;
+import java.util.function.Function;
+
+public interface ConfigurationProvider<T> extends AutoCloseable {
+
+    T get();
+    <C> ConfigurationProvider<C> createConfigurationProvider(Function<T, C> configurationExtractor);
+    void registerChangeListener(BiConsumer<T, T> listener);
```

Comment:
I can't guess based on it getting T, T which argument it receives.
Better have an interface for that.